src/benchq/compilation/graph_sim_mini.jl

```diff
@@ -29,228 +36,369 @@ Raises:
-        graph state
+    Vector{LCO}: the list of local clifford operations on each node
+    Vector{AdjList}:   the adjacency list describing the graph corresponding to the graph state
 """
-function get_graph_state_data(icm_circuit, n_qubits)
-    lco = [H_code for _ in 1:n_qubits]  # local clifford operation on each node
-    adj = [Set{Int}() for _ in 1:n_qubits]  # adjacency list
+function get_graph_state_data(icm_circuit::Vector{ICMOp}, n_qubits, display=false)
+    lco = fill(H_code, n_qubits)   # local clifford operation on each node
+    adj = [AdjList() for _ in 1:n_qubits]  # adjacency list
 
-    # for keeping track of progress
     total_length = length(icm_circuit)
-    last_10_percent_completed = 0
-    i = 0
-
-    for (op_name, op_qubits) in icm_circuit
-        i += 1
-        percent_completed = round(100 * i / total_length)
-        if percent_completed >= last_10_percent_completed + 10
-            println("GraphSim Mini is $percent_completed% completed")
-            last_10_percent_completed += 10
+    counter = 0
+    start_time = time()
+    erase = "        \b\b\b\b\b\b\b\b"
+
+    for icm_op in icm_circuit
+        if display
+            counter += 1
+            if counter > 1000
+                percent = round(Int, 100 * counter / total_length)
+                elapsed = round(time() - start_time, digits=2)
+                print("\r$(percent)% completed in $erase$(elapsed)s")
+                counter = 0
+            end
         end
 
-        qubit_1 = parse(Int, op_qubits[1]) + 1
-        if op_name in ["I", "X", "Y", "Z"]
-            # these gates do not change the graph
-            continue
-        elseif op_name == "H"
-            lco[qubit_1] = multiply_lco[H_code, lco[qubit_1]]
-        elseif op_name == "S"
-            lco[qubit_1] = multiply_lco[S_code, lco[qubit_1]]
-        elseif op_name == "S_Dagger"
-            lco[qubit_1] = multiply_lco[S_Dagger_code, lco[qubit_1]]
-        elseif op_name == "CZ"
-            cz(lco, adj, qubit_1, parse(Int, op_qubits[2]) + 1)
-        elseif op_name == "CNOT"
-            # CNOT = (I \otimes H) CZ (I \otimes H)
-            qubit_2 = parse(Int, op_qubits[2]) + 1
-            lco[qubit_2] = multiply_lco[H_code, lco[qubit_2]]
+        op_code = icm_op.code
+        qubit_1 = icm_op.qubit1
+        if op_code == H_code
+            lco[qubit_1] = multiply_h(lco[qubit_1])
+        elseif op_code == S_code
+            lco[qubit_1] = multiply_s(lco[qubit_1])
+        elseif op_code == CNOT_code
+            # CNOT = (I ⊗ H) CZ (I ⊗ H)
+            qubit_2 = icm_op.qubit2
+            lco[qubit_2] = multiply_h(lco[qubit_2])
             cz(lco, adj, qubit_1, qubit_2)
-            lco[qubit_2] = multiply_lco[H_code, lco[qubit_2]]
-        else
-            error("Unknown gate: $op_name")
+            lco[qubit_2] = multiply_h(lco[qubit_2])
+        else # CZ_code
+            cz(lco, adj, qubit_1, icm_op.qubit2)
```

Comment:
Shouldn't we have a case to catch when an unrecognized gate is encountered? If not, and we always expect `get_icm` to be run first, we should make this function private right?